src/dotnet/Db/DbSetExt.cs

```diff
@@ -26,4 +26,199 @@ public static class DbSetExt
             throw new ArgumentOutOfRangeException(nameof(key), "Key must be greater than zero.");
         return set.FindAsync(DbKey.Compose(key), cancellationToken);
     }
+
+    public static async Task Lock<TEntity, TKey>(
+        this DbSet<TEntity> set,
+        TKey key,
+        CancellationToken cancellationToken)
+        where TEntity : class
+    {
+        var context = set.GetDbContext();
+        var keyHash = (long)(key?.GetHashCode() ?? 239);
+        var lockKey = (keyHash << 32) ^ HashCode.Combine(typeof(TEntity).GetHashCode());
+        var timeout = context.Database.GetCommandTimeout();
+        // TODO(AK): find a way to resolve DBSettings instance
+        context.Database.SetCommandTimeout(30);
+        try {
+            await context.Database.ExecuteSqlAsync($"SELECT pg_advisory_xact_lock({lockKey});", cancellationToken);
+        }
+        catch (Exception ex) {
+            // TODO(AK): find a way to resolve a logger
+            await Console.Error.WriteLineAsync("Lock failed: " + ex.Message);
+            throw;
+        }
+        finally {
+            context.Database.SetCommandTimeout(timeout);
+        }
+    }
+
+    public static async Task Lock<TEntity, TKey, TArg0>(
+        this DbSet<TEntity> set,
+        TKey key,
+        TArg0 arg0,
+        CancellationToken cancellationToken)
+        where TEntity : class
+    {
+        var context = set.GetDbContext();
+        var keyHash = (long)(key?.GetHashCode() ?? 239);
+        var lockKey = (keyHash << 32) ^ HashCode.Combine(typeof(TEntity).GetHashCode(), arg0);
+        var timeout = context.Database.GetCommandTimeout();
+        try {
```

Comment:
It seems all Lock methods constists of 2 parts: lockKey calculation and  executing lock command. Extract executing lock command to a separate method and reuse it.